site/_posts/2018-11-19-r-spark-improvements.md

```diff
@@ -0,0 +1,153 @@
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+{% endcomment %}
+-->
+
+*[Javier Luraschi][1] is a software engineer at [RStudio][2]*
+
+Support for Apache Arrow in Apache Spark with R is currently under
+active development through [sparklyr][3]. This post explores early, yet
+promising, performance improvements achieved when using R with [Apache
+Spark][4] and Arrow.
+
+# Setup
+
+Since this work is under active development, install `sparklyr` and
+`arrow` from GitHub as follows:
+
+```r
+devtools::install_github("apache/arrow", subdir = "r", ref = "dc5df8f")
+devtools::install_github("rstudio/sparklyr")
+```
+
+In this benchmark, we will use [dplyr][5], but similar improvements can
+be  expected from using [DBI][6], or [Spark DataFrames][7] in `sparklyr`. 
+The local Spark connection and dataframe with 1M numeric rows was
+initialized as follows:
+
+```r
+library(sparklyr)
+library(dplyr)
+
+sc <- spark_connect(master = "local")
+data <- data.frame(y = runif(10^6, 0, 1))
+```
+
+# Copying
+
+The following benchmark using [microbenchmark][8], copies 1M rows from
+R into Spark using `sparklyr` with and without `arrow`, there is close
+to a 10x improvement using `arrow`.
+
+
+```r
+microbenchmark::microbenchmark(
+  setup = library(arrow),
+  arrow_on = {
+    library(arrow)
+    sparklyr_df <<- copy_to(sc, data, overwrite = T)
+    count(sparklyr_df)
+  },
+  arrow_off = {
+    if ("arrow" %in% .packages()) detach("package:arrow")
+    sparklyr_df <<- copy_to(sc, data, overwrite = T)
+    count(sparklyr_df)
+  },
+  times = 10
+) %>% ggplot2::autoplot()
+```
+
+<div align="center">
+<img src="{{ site.base-url }}/img/arrow-r-spark-copying.png"
+     alt="Copying data with R into Spark with and without Arrow"
+     width="60%" class="img-responsive">
+</div>
+
+# Collecting
+
+The following benchmark collects 1M rows from Spark into R and shows that `arrow`
+can bring 2x improvements. The collection improvements are not as significant as
+copying data since, `sparklyr` already collects data in columnar format.
+
+```r
+microbenchmark::microbenchmark(
+  setup = library(arrow),
+  arrow_on = {
+    dplyr::collect(sparklyr_df)
+  },
+  arrow_off = {
+    if ("arrow" %in% .packages()) detach("package:arrow")
+    dplyr::collect(sparklyr_df)
+  },
+  times = 10
+) %>% ggplot2::autoplot()
+```
+
+<div align="center">
+<img src="{{ site.base-url }}/img/arrow-r-spark-collecting.png"
+     alt="Collecting data with R from Spark with and without Arrow"
+     width="60%" class="img-responsive">
+</div>
+
+# Transforming
+
+Custom transformations of data using R functions are about 100X faster using `arrow`.
```

Comment:
How are custom transformations applied, does every spark node forks a R process and copy to the forked process then gather the result (thus implying 2 copy per partition)?

I'd say this would be relevant information for someone like me who has no knowledge on how sparklyr works.